### PR TITLE
Feature weakref module

### PR DESCRIPTION
**Pull request by vallsv**
- Create silx.utils module
- Create a WeakMethod (class with callback)
- Create a WeakList

### CODE REVIEW

File: silx/utils/weakref.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+            list
+        """
+        list.__init__(self)
+        self.__list = []
+        self.__is_valid = True
+        for obj in enumerator:
+            self.append(obj)
+
+    def __invalidate(self, ref):
+        """Flag the list as invalidated. The contains dead references."""
```
Comment (dnaudet) [inline silx/utils/weakref.py]: the \***list**

File: silx/utils/weakref.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+__authors__ = ["V. Valls"]
+__license__ = "MIT"
+__date__ = "24/08/2016"
+
+
+import weakref
+import types
+import inspect
+from ..resources import resource_filename
```
Comment (dnaudet) [inline silx/utils/weakref.py]: ?

**Comment by dnaudet** [comment]
How about a WeakMethodProxy? Inherit from WeakMethod and override __call__.

**Comment by vallsv** [comment]
- The absolute import is needed to force the module to load the lib saferef, else it tries to load itself :-D
- As you want for the proxy, but my feeling with proxy is it is more difficult to manage
- I dont understand your comment about list. I see a typo on the comment is it that?

Let me clean up a little more this request. I will try to have a full coverage of the code and to implement the sort function in a simple way.

**Comment by dnaudet** [comment]
* no i meant *from ..resources import resource_filename* :wink: 
* see below
* yes i supposed the doc meant to say : The **list** contains dead references.

For the proxy, something along the lines of (there's probably a pitfall i didn't think of tho...) :
```python
class WeakMethodProxy(WeakMethod):
    def __call__(self, *args, **kwargs):
        fn = super(WeakMethodProxy, self).__call__()
        if fn is None:
            raise something
        return fn(*args, **kwargs)
```

**Comment by vallsv** [comment]
Here it is ready to merge.

File: silx/utils/weakref.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+class WeakList(list):
+    """Manage a list of weaked references.
+    When an object is dead, the list is flaged as invalid.
+    If expected the list is cleaned up to remove dead objects.
+    """
+
+    def __init__(self, enumerator=()):
+        """Create a WeakList
+
+        :param enumerator list or tuple: A list of object to initialize the
```
Comment (t20100) [inline silx/utils/weakref.py]: syntax: :param [type] variable_name:
-> :param iterable enumerator:

File: silx/utils/weakref.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+    def reverse(self):
+        """Reverse the list inplace"""
+        self.__list.reverse()
+
+    def sort(self, cmp=None, key=None, reverse=False):
+        """Sort the list inplace.
+        Not very efficient."""
+        sorted_list = list(self).sort(cmp, key, reverse)
+        return WeakList(sorted_list)
```
Comment (t20100) [inline silx/utils/weakref.py]: I think it should be: self.__list.sort() for inplace sort (as for reverse)

File: silx/utils/weakref.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+    If expected the list is cleaned up to remove dead objects.
+    """
+
+    def __init__(self, enumerator=()):
+        """Create a WeakList
+
+        :param enumerator list or tuple: A list of object to initialize the
+            list
+        """
+        list.__init__(self)
```
Comment (t20100) [inline silx/utils/weakref.py]: It seems list inheritance is not used as all methods uses self.__list.
Is it needed?

File: silx/utils/weakref.py (hunk 3 above)
Comment (vallsv) [inline silx/utils/weakref.py]: Hooo shit, i do the reverse thing from a long time...

File: silx/utils/weakref.py (hunk 4 above)
Comment (vallsv) [inline silx/utils/weakref.py]: It will test cmp on references objects, i dont know if it is part of the ref API

File: silx/utils/weakref.py (hunk 5 above)
Comment (vallsv) [inline silx/utils/weakref.py]: You mean if `__list` is needed? Yes you are right, maybe no.
But it is better to inherite from list for that: `isisntance(weaklist, list)`.
I can try to remove `__list`.

**Comment by vallsv** [comment]
Here is a fix of the sort method, and a fix on the param docstring.
Ready to merge.

**Pull request merged**